scripts/Diffraction/single_crystal/sxd.py

```diff
@@ -0,0 +1,320 @@
+                               <radius val="0.003"/>
+                               </sphere>"""  # sphere radius 3mm  - used for vanadium and NaCl
+
+    def process_data(self, runs: Sequence[str], *args):
+        gonio_angles = args
+        if len(gonio_angles) != len(self.gonio_axes):
+            return
+        for irun, run in enumerate(runs):
+            wsname = self.load_run(run)
+            # set goniometer
+            if self.gonio_axes is not None:
+                # gonio_angles are a list of motor strings (same for all runs)
+                if isinstance(gonio_angles[0], str):
+                    self._set_goniometer_on_ws(wsname, gonio_angles)
+                else:
+                    # gonio_angles is a list of individual or tuple motor angles for each run
+                    self._set_goniometer_on_ws(wsname, gonio_angles[irun])
+            # normalise by vanadium
+            self._divide_workspaces(wsname, self.van_ws)
+            # set sample (must be done after gonio to rotate shape) and correct for attenuation
+            if self.sample_dict is not None:
+                mantid.SetSample(wsname, EnableLogging=False, **self.sample_dict)
+                mantid.ConvertUnits(InputWorkspace=wsname, OutputWorkspace=wsname, Target="Wavelength", EnableLogging=False)
+                if "<sphere" in ADS.retrieve(wsname).sample().getShape().getShapeXML():
+                    transmission = mantid.SphericalAbsorption(InputWorkspace=wsname, OutputWorkspace="transmission", EnableLogging=False)
+                else:
+                    transmission = mantid.MonteCarloAbsorption(
+                        InputWorkspace=wsname, OutputWorkspace="transmission", EventsPerPoint=self.n_mcevents, EnableLogging=False
+                    )
+                self._divide_workspaces(wsname, transmission)
+                mantid.DeleteWorkspace(transmission)
+            # save results in dictionary
+            self.set_ws(run, wsname)
+        return wsname
+
+    def load_run(self, runno):
+        wsname = "SXD" + str(runno)
+        mantid.Load(Filename=wsname + ".raw", OutputWorkspace=wsname, EnableLogging=False)
+        if self.detcal_path is not None:
+            mantid.LoadParameterFile(Workspace=wsname, Filename=self.detcal_path, EnableLogging=False)
+        mantid.CropWorkspace(InputWorkspace=wsname, OutputWorkspace=wsname, XMin=700, XMax=18800, EnableLogging=False)
+        mantid.NormaliseByCurrent(InputWorkspace=wsname, OutputWorkspace=wsname, EnableLogging=False)
+        return wsname
+
+    def _replace_spectra_with_focussed_bank(self, wsname):
+        # important to focus in TOF not d-spacing (lambda would be better but not compatible with Diffractionfocussing)!
+        ws = SXD.retrieve(wsname)
+        banks_foc = mantid.GroupDetectors(
+            InputWorkspace=ws,
+            OutputWorkspace="grouped",
+            IgnoreGroupNumber=False,
+            CopyGroupingFromWorkspace=self.grp_ws,
```

Comment:
Perhaps the grp_ws could just be passed into this function as a parameter. Doesn't seem like it's used anywhere else so having as a field on the class may be unnecessary